Spatial_sorting/include/CGAL/Hilbert_sort_median_3.h

```diff
@@ -114,16 +121,96 @@ class Hilbert_sort_median_3
         RandomAccessIterator m5 = internal::hilbert_split (m4, m6, Cmp< z,  upz> (_k));
         RandomAccessIterator m7 = internal::hilbert_split (m6, m8, Cmp< z, !upz> (_k));
 
-        sort<z, upz, upx, upy> (m0, m1);
-        sort<y, upy, upz, upx> (m1, m2);
-        sort<y, upy, upz, upx> (m2, m3);
-        sort<x, upx,!upy,!upz> (m3, m4);
-        sort<x, upx,!upy,!upz> (m4, m5);
-        sort<y,!upy, upz,!upx> (m5, m6);
-        sort<y,!upy, upz,!upx> (m6, m7);
-        sort<z,!upz,!upx, upy> (m7, m8);
+        recursive_sort<z, upz, upx, upy> (m0, m1);
+        recursive_sort<y, upy, upz, upx> (m1, m2);
+        recursive_sort<y, upy, upz, upx> (m2, m3);
+        recursive_sort<x, upx,!upy,!upz> (m3, m4);
+        recursive_sort<x, upx,!upy,!upz> (m4, m5);
+        recursive_sort<y,!upy, upz,!upx> (m5, m6);
+        recursive_sort<y,!upy, upz,!upx> (m6, m7);
+        recursive_sort<z,!upz,!upx, upy> (m7, m8);
     }
 
+  template <int x, bool upx, bool upy, bool upz, class RandomAccessIterator>
+  struct Recursive_sort {
+    const Self& hs;
+    RandomAccessIterator begin,end;
+    
+    Recursive_sort(const Self& hs, RandomAccessIterator begin, RandomAccessIterator end)
+      : hs(hs), begin(begin), end(end)
+    {}
+    
+    void operator()() const
+    {
+      hs.recursive_sort<x,upx,upy,upz>(begin,end);
+    }
+  };
+  
+  template <class RandomAccessIterator, class Comp>
+  struct Hilbert_split
+  {
+    RandomAccessIterator& res;
+    RandomAccessIterator begin, end;
+    const Comp& comp;
+    
+    Hilbert_split(RandomAccessIterator& res, RandomAccessIterator begin, RandomAccessIterator end, const Comp comp)
+      : res(res), begin(begin), end(end), comp(comp)
+    {}
+    
+    void operator()() const
+    {
+      res = internal::hilbert_split(begin, end, comp);
+    }
+  };
+  
+    
+  template <int x, bool upx, bool upy, bool upz, class RandomAccessIterator>
+  void sort (RandomAccessIterator begin, RandomAccessIterator end) const
+  {
+    const int y = (x + 1) % 3, z = (x + 2) % 3;
+    if ((end - begin) <= _limit) return;
+
+    RandomAccessIterator m0 = begin, m8 = end;
+#if defined( CGAL_LINKED_WITH_TBB) && ( ! defined (CGAL_NO_PARALLEL_SPATIAL_SORT) )
+    if((end - begin) > 1024){
+      RandomAccessIterator m1, m2, m3, m4, m5, m6, m7;
+      m4 = internal::hilbert_split(m0, m8, Cmp<x, upx>(_k));
+      {
+        tbb::task_group g;
+        g.run(Hilbert_split<RandomAccessIterator,Cmp<y, upy> >(m2,m0,m4,Cmp<y,  upy>(_k)));
+        g.run(Hilbert_split<RandomAccessIterator,Cmp<y, !upy> >(m6,m4,m8,Cmp<y,  !upy>(_k)));
+        g.wait();
```

Comment:
I think the last run can be merged with wait into run_and_wait. But I think Intel recommends parallel_invoke instead to execute a fixed number of functions in parallel. Did you find that task groups are more efficient or more convenient?

Reply:
It's true that `parallel_invoke()` reads more nicely. The speedup is the same.